File: apps/ios/nearby/Features/Send/RecipientViewModel.swift
import Combine
import Foundation
import UI

/// Drives the recipient field's edge-of-field state machine: idle → resolving → resolved ✓ / not-found
/// ✗. Addresses resolve instantly; SuiNS names debounce then hit the network.
@MainActor
final class RecipientViewModel: ObservableObject {
  enum State: Equatable {
    case idle
    case invalid
    case resolving
    case resolved(address: String, name: String?)
    case notFound
  }

  @Published
  private(set) var input = ""

  @Published
  private(set) var state: State = .idle

  private let service: RecipientService
  private let toastController: ToastController?

  private var resolveTask: Task<Void, Never>?

  init(
    service: RecipientService = RecipientService(network: AppConstants.suiNetwork),
    toastController: ToastController? = nil
  ) {
    self.service = service
    self.toastController = toastController
  }

  /// The address to send to, once a recipient resolves.
  var resolvedAddress: String? {
    if case .resolved(let address, _) = state { return address }
    return nil
  }

  /// Entry point from the text field: reparse, then resolve names after a debounce.
  func onInputChange(_ raw: String) {
    input = raw
    resolveTask?.cancel()

    switch RecipientInput.parse(raw) {
    case .empty:
      state = .idle
    case .invalid:
      state = .invalid
    case .address(let address):
      state = .resolved(address: address, name: nil)
    case .name(let name):
      state = .resolving
      resolveTask = Task { await resolve(name: name) }
    }
  }

  private func resolve(name: String) async {
    try? await Task.sleep(nanoseconds: UInt64(AppConstants.nameCheckDebounce * 1_000_000_000))
    guard !Task.isCancelled else { return }
    let svc = service
    do {
      let address = try await withTimeout(seconds: AppConstants.networkTimeout) {
        try await svc.resolve(name: name)
      }
      guard !Task.isCancelled else { return }
      state = address.map { .resolved(address: $0, name: name) } ?? .notFound
    } catch is TimeoutError {
      guard !Task.isCancelled else { return }
      state = .notFound
      toastController?.show("Name lookup timed out. Check your connection.", tone: .warning)
    } catch {
      guard !Task.isCancelled else { return }
      state = .notFound
    }
  }
}

File: apps/android/app/src/main/java/com/variance/nearby/screens/send/RecipientViewModel.kt
package com.variance.nearby.screens.send

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.variance.nearby.core.AppConstants
import com.variance.nearby.services.sui.RecipientService
import com.variance.nearby.ui.ToastController
import com.variance.nearby.ui.ToastTone
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import kotlin.time.Duration.Companion.milliseconds

/**
 * Drives the recipient field's edge-of-field state machine: Idle → Resolving → Resolved ✓ / NotFound
 * ✗. Addresses resolve instantly; SuiNS names debounce then hit the network.
 */
class RecipientViewModel(
    private val service: RecipientService,
    private val toastController: ToastController? = null,
) : ViewModel() {

    sealed interface State {
        data object Idle : State
        data object Invalid : State
        data object Resolving : State
        data class Resolved(val address: String, val name: String?) : State
        data object NotFound : State
    }

    var input by mutableStateOf("")
        private set

    var state by mutableStateOf<State>(State.Idle)
        private set

    private var resolveJob: Job? = null

    /** The address to send to, once a recipient resolves. */
    val resolvedAddress: String?
        get() = (state as? State.Resolved)?.address

    /** Entry point from the text field: reparse, then resolve names after a debounce. */
    fun onInputChange(raw: String) {
        input = raw
        resolveJob?.cancel()

        when (val parsed = RecipientInput.parse(raw)) {
            RecipientInput.Empty -> state = State.Idle
            RecipientInput.Invalid -> state = State.Invalid
            is RecipientInput.Address -> state = State.Resolved(parsed.value, null)
            is RecipientInput.Name -> {
                state = State.Resolving
                resolveJob = viewModelScope.launch {
                    delay(AppConstants.NAME_CHECK_DEBOUNCE_MS.milliseconds)
                    try {
                        val address = withTimeout(AppConstants.NETWORK_TIMEOUT_MS) {
                            withContext(Dispatchers.IO) { service.resolve(parsed.value) }
                        }
                        state = if (address != null) {
                            State.Resolved(address, parsed.value)
                        } else {
                            State.NotFound
                        }
                    } catch (e: TimeoutCancellationException) {
                        state = State.NotFound
                        toastController?.show(
                            "Name lookup timed out. Check your connection.",
                            ToastTone.WARNING,
                        )
                    } catch (_: Exception) {
                        state = State.NotFound
                    }
                }
            }
        }
    }
}
